### ISSUE TITLE
Regression in resizing performance

### ISSUE THREAD

**Issue by clo4** [zed-industries/community#2405]
### Check for existing issues

- [X] Completed

### Describe the regression

There's a noticeable regression in the performance in resizing the window. In the gpui1 build, resizing at all window sizes feels instant and responsive, but in 2, it's sluggish at larger window sizes. I've attached a video showing this, and you *can* see it, but it's harder to see than it is to feel, so try it yourself if you can. This may be down to specs (I'm on M1) -- it's flawless in the OG version so it does feel like a regression, but I get it if this is a wontfix :P

### If necessary, add screenshots of gpui-1 Zed to show the missing feature

[Zed2 resizing.webm](https://github.com/zed-industries/community/assets/52195359/bfd17028-01c0-4759-b4a4-82a9c244e651)

**Comment by clo4** [comment]
Hah, video was compressed to 30fps making it basically impossible to see. Guess you'll have to take my word for it!

**Comment by JosephTLyons** [comment]
Thanks a ton @clo4, I will highlight this one in tomorrows meeting.

**Comment by JosephTLyons** [comment]
Are you able to host the video in another way @clo4, I'm having a hard time reproducing this myself.

**Comment by clo4** [comment]
Done, I made it unlisted on YouTube so it's only visible to this thread. https://youtu.be/5rmJTJyVgoA

This is still happening in the latest beta and it's very reproducible for me, let me know if you'd like me to run any tests or get any logs/diagnostics

**Comment by ConradIrwin** [comment]
Playing around with this today and some of our performance tooling, it seems like we're doing an "ok but not perfect" job of rendering during resizes (at least on my machine):

<img width="1107" alt="Screenshot 2024-01-22 at 4 05 08 PM" src="https://github.com/zed-industries/community/assets/94272/8b4932ba-7455-43fd-8c4d-7de3ae8939c4">

If you want to try and reproduce this graph on your end:

* Run the latest preview with `ZED_MEASUREMENTS=1` 
* Dump the output to a file (say `zed.measure`)
* Run `./script/histogram zed.measure` to convert the output into the above (or upload the log here and I can run the script).

I'd be curious to know what this is like on your machine, as I don't notice much of a qualitative difference between stable zed and preview.

That said, it's not totally clear whether this number accurately correlates to "feeling smooth". In particular if I record the screen while doing this, I notice that the dragged corner of the app is usually a frame (or more) behind the cursor position ; so it's possible that we're already on the back foot before hitting these times (though I observe the same with stable Zed, and with finder rendering an empty folder, so that may just be "how it is").

Adding more logging I can see that the resize handler is being called pretty reliably every 16ms by the operating system while I'm actively resizing (there are a couple of drops, but I'm not sure if that's because my mouse was temporarily still, or some other artifact).

If I add a new measurement, from the time that we receive the resize event to when we finish drawing the frame then we can see something more (potentially) interesting:

<img width="1312" alt="Screenshot 2024-01-22 at 4 43 08 PM" src="https://github.com/zed-industries/community/assets/94272/10e4b071-f14a-462b-92f6-967612abc025">

I say potentially interesting because I think this means that mostly it takes a full 16ms (or a frame-size) from receiving the scroll event to finishing rendering the next frame. That said, I *think* this may be the case because we receive the resize immediately after drawing, and we don't draw another frame until requested. That does reveal a few % of cases where we likely are missing the "one update per frame" window, but that doesn't seem sufficient to account for what you're talking about.

**Comment by clo4** [comment]
I don't have access to the histogram script but here's the performance numbers for you:
[zed_preview.txt](https://github.com/zed-industries/community/files/14018719/zed_preview.txt)

Note that this is with a file open. When no file is open, or a file without highlighting, Zed hits the frame target very consistently during resize. That makes me think it may be due to syntax highlighting?

---

Also, just to check, is it expected that the frame duration for mousing over the file browser tends to be around 30ms? That seems quite high for a relatively small change, but I don't have any numbers to compare it to.

**Comment by clo4** [comment]
Oh- weird. When I have a file with highlighting open, the frame durations are noticeably higher across the whole app.

**Comment by as-cii** [comment]
Thanks so much for those measurements @clo4. Those frame times don't look great and it's interesting that the slowdown shows up when you have a file open.

Do you have macOS `Instruments.app` installed on your machine? If so, would you mind capturing a `CPU Profiler` while you resize? That would be really helpful to understand where time is going given that we can't reproduce this on our machines.

You can export the recorded profile and attach it to this thread so that we can dig into it. Thanks again! 🙏

**Comment by clo4** [comment]
Huh. In capturing the trace, I think I figured it out. I had a font set as my `buffer_font_family` but after I set up my new system I kept the same config but didn't reinstall the font. It looks (to me) like on every redraw Zed looks up the font and that fails -- seems like an expensive operation. When I set my buffer font to a font that *is* installed on my machine, no performance issues at all. I've attached the trace anyway because it's not unlikely that I'm wrong, lol. Zipped because of GitHub.
[Zed Preview Resizing.trace.zip](https://github.com/zed-industries/community/files/14024883/Zed.Preview.Resizing.trace.zip)
